python/tvm/relay/frontend/pytorch.py

```diff
@@ -0,0 +1,1026 @@
+        data = inputs[0]
+
+        import torch
+        if isinstance(data, _expr.Expr):
+            shape = _infer_shape(data)
+        elif isinstance(data, list):
+            shape = data
+        elif isinstance(data, (torch.Tensor, np.ndarray)):
+            shape = data.shape
+        else:
+            assert "data type {} could not be parsed in ones op" % (type(data))
+
+        return _op.full(_expr.const(1), shape, dtype=_convert_data_type(input_types[0]))
+    return _impl
+
+def _zeros():
+    def _impl(inputs, input_types):
+        data = inputs[0]
+
+        import torch
+        if isinstance(data, _expr.Expr):
+            shape = _infer_shape(data)
+        elif isinstance(data, list):
+            shape = data
+        elif isinstance(data, (torch.Tensor, np.ndarray)):
+            shape = data.shape
+        else:
+            assert "data type {} could not be parsed in zeros op" % (type(data))
+
+        return _op.full(_expr.const(0), shape, dtype=_convert_data_type(input_types[0]))
+    return _impl
+
+def _relu():
+    def _impl(inputs, input_types):
+        data = inputs[0]
+        return _op.nn.relu(data)
+    return _impl
+
+def _adaptive_avg_2d():
+    def _impl(inputs, input_types):
+        data = inputs[0]
+        output_size = _infer_shape(inputs[1])
+
+        return _op.contrib.contrib.adaptive_avg_pool2d(
+            data,
+            output_size=output_size)
+    return _impl
+
+def _adaptive_max_2d():
+    def _impl(inputs, input_types):
+        data = inputs[0]
+        output_size = _infer_shape(inputs[1])
+
+        return _op.contrib.contrib.adaptive_max_pool2d(
+            data,
+            output_size=output_size)
+    return _impl
+
+def _maxpool_2d():
+    def _impl(inputs, input_types):
+        data = inputs[0]
+
+        pool_size = _infer_shape(inputs[1])
+        strides = _infer_shape(inputs[2])
+        padding = _infer_shape(inputs[3])
+
+        ceil_mode = int(inputs[5])
+
+        return _op.nn.max_pool2d(data, pool_size, strides, padding, "NCHW", ceil_mode)
+    return _impl
+
+def _hardtanh():
+    def _impl(inputs, input_types):
+        a = inputs[0]
+        tanh_min = float(inputs[1])
+        tanh_max = float(inputs[2])
+        return _op.tensor.clip(a, tanh_min, tanh_max)
+    return _impl
+
+def _convolution():
+    def _impl(inputs, input_types):
+        # Use transpose or normal
+        use_transpose = True if inputs[6] == "1" else False
+
+        data = inputs[0]
+        weight = inputs[1]
+        bias = inputs[2]
+        strides = inputs[3]
+        padding = inputs[4]
+        dilation = inputs[5]
+
+        if isinstance(weight, _expr.Expr):
+            inferred_shape = _infer_shape(weight)
+            weight_shape = []
+            for infer in inferred_shape:
+                weight_shape.append(infer)
+        else:
+            assert "data type {} could not be parsed in conv op" % (type(weight))
+
+        channels = weight_shape[0]
```

Comment:
This could be wrong for depthwise conv2d. `channels` should be the number of channels in output. For depthwise conv2d, output channels should be input channels * channel multiplier, which is `weight_shape[0] * weight_shape[1]`. Need a special handling here.

Reply:
This is correct for torch. Users are expected to provide the right output channels. For depth wise conv, `weight_shape[0]` is equal to `groups`.

See their mobilenet v2 implementation
https://github.com/pytorch/vision/blob/1e27b533490d5b375c4656d2dd67961baaa1dced/torchvision/models/mobilenet.py#L121

Reply:
If that's the case, you need to reshape the weight into `(in_channel, channel_multiplier, kernel_h, kernel_w)` as it's the expected shape for `depthwise_conv2d` in relay.

Reply:
It is already in that shape. We can run mobilenet v2 from torchvision and get the correct output. If I dump the weight shapes there, one of them is `[576, 1, 3, 3]` and there is a autotvm logging 

```
Cannot find config for target=llvm, workload=('depthwise_conv2d_nchw', (1, 576, 14, 14, 'float32'), (576, 1, 3, 3, 'float32'), (1, 1), (1, 1, 1, 1), (1, 1), 'float32'). A fallback configuration is used, which may bring great performance regression.
```

Reply:
How about the situation when channel multiplier > 1?

Reply:
I think torch users are responsible for taking care of channel multiplier. Again, this is an example from torch mobilenet v2 implementation where there is an multiplication by `width_mult`
https://github.com/pytorch/vision/blob/1e27b533490d5b375c4656d2dd67961baaa1dced/torchvision/models/mobilenet.py#L121

Reply:
I'm saying that we need to convert depthwise conv2d correctly to relay depthwise conv2d semantics. When channel multiplier > 1, output channels = channel_multiplier * input_channels. Further, we should look at pytorch [API definition](https://pytorch.org/docs/stable/nn.html#conv2d) instead of mobilenet implementation.  I don't know how pytorch defines the weight shape.

If weight shape is `[out_channels, 1, kernel_h, kernel_w]`, we then need to reshape it to `[input_channels, channel_multiplier, kernel_h, kernel_w]`.

If weight shape is `[in_channels, channel_multiplier, kernel_h, kernel_w]`, then here `channels = weight_shape[0] * weight_shape[1]`.

Does it make sense?

Reply:
Yes, it does make sense now that I did some research on this topic.  I now understand why we need reshape when channel_multipler > 1. Let me explain my findings:

First, a background on torch:
* According to their [API doc](https://pytorch.org/docs/stable/nn.html#conv2d), it says 
```
When groups == in_channels and out_channels == K * in_channels, 

where K is a positive integer,  this operation is also termed in literature as depthwise convolution.
```

* In torch, weight_shape[1] is always in_channel // groups ([source](https://github.com/pytorch/pytorch/blob/00c6b903278d574d94171ebbf01986d698e22716/torch/nn/modules/conv.py#L46-L47)). So for depth wise conv, weight_shape[1] is always 1.

However in Relay, depth wise conv occurs when the **output** channel is equal to `groups`. See https://github.com/apache/incubator-tvm/blob/2c0c18494e00dd3beb71527b3f2ccde5df678440/src/relay/op/nn/convolution.h#L164-L168

So in the current implementation, if channel multiplier > 1, this condition is always false and depth wise convolution is never used even though it can. But the output is still correct, because we are doing regular convolution and comparing its output with torch's depth wise convolution output.

For depth wise convolution to happen correctly, I agree with @icemelon9 that we should add reshape when channel multiplier > 1.

@icemelon9 Does this sound about right? I also wonder if other frontend are handling depth wise conv + mulitplier > 1 case correctly.

Reply:
@icemelon9 can you also confirm if the condition `tvm::tir::Equal(param->channels, param->groups)` above is correct? This is different from torch semantics but torch API doc says that torch's terminology is the standard in the literature.

Reply:
@masahi Thanks for looking into this. You're right. Currently the relay type rel for depthwise conv2d is wrong. I've fixed it when developing the relay op strategy #4644 (see [here](https://github.com/apache/incubator-tvm/pull/4644/files#diff-d16c1ec50ceaac0e06f963a425d325b4R156-R165)). 

Because it involves the change to both relay and topi, and since the op strategy PR is close to be merged, we can first make sure the frontend converter is correct in this PR and later we can add a test case for channel multiplier > 1.

Reply:
I expect your PR will merged sooner than this one. We can add a multiplier > 1 test here after rebase.

Reply:
Thanks for the discussion and a fix.

Reply:
Since #4644 is merged, we can now fix the issue when channel multiplier > 1 and add the test in this pr.

Reply:
I can see why multiplying groups by channel multiplier makes sense but it seems to run into a shape mismatch when type checking the conv op. Just passing in groups doesn't run into the same issue which I'm trying to understand. With the change, we are actually using depth wise conv when channel multiplier > 1 now though.

Reply:
Have you rebased the PR against the latest master? It should have fixed the bug in relay type relation function.

Reply:
Yes, I'm currently up to date and have #4644 changes here.

Reply:
@alexwong that was a workaround for the relay bug, now that it is fixed you don't need to multiply the group

Reply:
Ah.. you're right. Was mistaken, didn't know that file was touched.